app/controllers/oaisys/pmh_controller.rb

```diff
@@ -28,7 +28,34 @@ def identify
   end
 
   def list_sets
-    expect_args exclusive: [:resumptionToken]
+    parameters = expect_args exclusive: [:resumptionToken]
+
+    resumption_token_provided = parameters[:page].present?
+    parameters[:page] = 1 if parameters[:page].blank?
+    sets_model, total_count, cursor = sets_on_page(page: parameters[:page])
+
+    if sets_model.out_of_range? && resumption_token_provided
+      raise Oaisys::BadResumptionTokenError.new, I18n.t('error_messages.resumption_token_invalid')
+    end
+
+    top_level_sets = Oaisys::Engine.config.top_level_sets_model.pluck(:id, :title)
+    sets = sets_model.pluck(:community_id, :id, :title, :description)
+
+    sets.map! do |top_level_sets_id, id, title, description|
+      [top_level_sets_id + ':' + id, top_level_sets.find { |a| a[0] == top_level_sets_id }[1] + ' / ' + title,
+       description]
```

Comment:
This should probably be split into a couple of lines with some temp variables just for general readability? It's a very complex one-liner.